wireguard-control/src/backends/userspace.rs

```diff
@@ -266,14 +266,27 @@ pub fn get_by_name(name: &InterfaceName) -> Result<Device, io::Error> {
 /// wgctrl-rs will look for WG_USERSPACE_IMPLEMENTATION first, but will also
 /// respect the WG_QUICK_USERSPACE_IMPLEMENTATION choice if the former isn't
 /// available.
-fn get_userspace_implementation() -> String {
-    std::env::var("WG_USERSPACE_IMPLEMENTATION")
-        .or_else(|_| std::env::var("WG_QUICK_USERSPACE_IMPLEMENTATION"))
-        .unwrap_or_else(|_| "wireguard-go".to_string())
+fn get_userspace_implementation() -> io::Result<PathBuf> {
+    let wg_path: PathBuf = PathBuf::from(
+        std::env::var("WG_USERSPACE_IMPLEMENTATION")
+            .or_else(|_| std::env::var("WG_QUICK_USERSPACE_IMPLEMENTATION"))
+            .unwrap_or_else(|_| "wireguard-go".to_string()),
+    );
+    if !wg_path.exists() {
```

Comment:
Ah yeah, I think we might want to rethink this as simply providing better output when we run the command in `start_userspace_wireguard()`. Personally I don't think the benefits of adding a dependency are worth it compared to adding a better "wireguard-go failed to execute - is it installed?" type message to the user.

Reply:
i *think* there is a switch for `--version` in `wireguard-go`. We could try to execute `wireguard-go --version`. If that succeeds, then `wireguard-go` exists in the path. 

```shell
  » /usr/bin/wireguard --version 
  wireguard-go v0.0.20220117
  
  Userspace WireGuard daemon for linux-amd64.
  Information available at https://www.wireguard.com.
  Copyright (C) Jason A. Donenfeld <Jason@zx2c4.com>.
  » echo $? 
  0
```

Reply:
@noyez I don't think that's necessary: https://play.rust-lang.org/?version=stable&mode=debug&edition=2021&gist=d93bf64dbf49743403648a4dfebf8b46

We can just check the result of `output()` and the result will be:

```rust
Err(Os { code: 2, kind: NotFound, message: "No such file or directory" })
```

when the binary is missing :).

So I think this PR is more about modifying `start_userspace_wireguard()` to provide more informative error messages up the chain for the user.

Reply:
So if the `output()` fails it could mean `wireguard-go` doesn't exists, or perhaps there's another error with the configurations. In `start_userspace_wireguard()` you already check if `output()` returns an error status, if so it returns `Err(io::ErrorKind::AddrNotAvailable.int())`.

Do you know to differentiate the different error codes?

Reply:
> Do you know to differentiate the different error codes?

Oh, `kind: NotFound` .... sorry, i missed that for some reason.

Reply:
How about something like the following? 

```
fn start_userspace_wireguard(iface: &InterfaceName) -> io::Result<Output> {
    let mut command = Command::new(get_userspace_implementation());

    let output_res = if cfg!(target_os = "linux") {
        command.args(&[iface.to_string()]).output()
    } else {
        command
            .env("WG_TUN_NAME_FILE", &format!("{VAR_RUN_PATH}/{iface}.name"))
            .args(["utun"])
            .output()
    };

    match output_res {
        Ok(output) => { 
            if output.status.success() { Ok(output) }
            else { Err(io::ErrorKind::AddrNotAvailable.into()) }
        }
        Err(e) => { 
            if e.kind() == io::ErrorKind::NotFound {
                Err(io::Error::new(
                    io::ErrorKind::NotFound,
                    &*format!("Cannot find \"wireguard-go\". Specify a custom path with WG_USERSPACE_IMPLEMENTATION.",)
                )).into()

            }
            else {
                Err(io::ErrorKind::AddrNotAvailable.into())
            }
        }
    }
}
```

Reply:
@noyez at first glance this looks reasonable, we can simplify it a bit and I think we don't want to return `io::ErrorKind::AddrNotAvailable` if we get an error running the command (no `output` was produced)

```rust
fn start_userspace_wireguard(iface: &InterfaceName) -> io::Result<Output> {
    let mut command = Command::new(get_userspace_implementation());

    let output_res = if cfg!(target_os = "linux") {
        command.args(&[iface.to_string()]).output()
    } else {
        command
            .env("WG_TUN_NAME_FILE", &format!("{VAR_RUN_PATH}/{iface}.name"))
            .args(["utun"])
            .output()
    };

    match output_res {
        Ok(output) => {
            if output.status.success() {
                Ok(output)
            } else {
                Err(io::ErrorKind::AddrNotAvailable.into())
            }
        },
        Err(e) if e.kind() == io::ErrorKind::NotFound => Err(io::Error::new(
            io::ErrorKind::NotFound,
            "Cannot find \"wireguard-go\". Specify a custom path with WG_USERSPACE_IMPLEMENTATION.",
        )),
        Err(e) => Err(e),
    }
}
```

Reply:
Yeah, I actually also don't get why we're returning `io::ErrorKind::AddrNotAvailable` for all errors, but I can fix that up in another PR since it's not related to wireguard-go being missing altogether.

Reply:
yeah, i didn't know what else to return, that's what was getting returned previously in the `start_userspace_wireguard` function so i kept it, but can def be changed.

Reply:
Yeah I think that one is my bad ;). It'd make this PR less "short and sweet" but I feel like maybe the right thing to do is to pass along the return code as well as possibly the `stderr` output so that we can output it to the user.

Feel free to take that on if you wish, otherwise I can address it in a later PR since it's my oversight.

Reply:
> I actually also don't get why we're returning io::ErrorKind::AddrNotAvailable for all errors

The previous code had a question mark operator:

https://github.com/tonarino/innernet/blob/376ab64ed2af5feb920573ab4877ab41b05d749c/wireguard-control/src/backends/userspace.rs#L278

So `ErrorKind::AddrNotAvailable` is only returned when there's not a success status code after running the command. But you're right that we should return all the details to the caller.